src/context/AuthContext.tsx

```diff
@@ -0,0 +1,80 @@
+'use client';
+
+import nookies from 'nookies';
+import React, {
+  createContext,
+  useState,
+  useContext,
+  useEffect,
+  useMemo,
+  ReactNode,
+} from 'react';
+
+import { getUser } from '../client';
+import { User } from '../types';
+import { SECONDS_IN_WEEK } from '../utils/constants';
+
+type AuthContextProps = {
+  user: User;
+  login: (token: string) => void;
+  logout: () => void;
+};
+
+const DefaultUser = {
+  id: -1,
+  name: '',
+  email: '',
+  rating: 0,
+};
+
+export const AuthContext = createContext<AuthContextProps | null>(null);
+
+type AuthProviderProps = {
+  children: ReactNode;
+};
+
+export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
+  const [user, setUser] = useState<User>(DefaultUser);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const { token } = nookies.get();
+      if (token) {
+        try {
+          const userData = await getUser(token);
+          setUser(userData);
+        } catch (error) {
+          nookies.destroy(null, 'token');
+        }
+      }
+    };
+
+    fetchData();
+  }, []);
```

Comment:
We could implement `useAsyncEffect` I like the clarity of it

```
import { useEffect } from 'react';

export const useEffectAsync = (fn: () => PromiseLike<any>, deps?: any[]) =>
  useEffect(() => {
    fn();
  }, deps);
```

add it to new /hooks folder

Reply:
pretty 🤗